teachertool/src/transforms/loadToolboxCategoriesAsync.ts

```diff
@@ -0,0 +1,33 @@
+import { getToolboxCategories } from "../services/makecodeEditorService";
+import { stateAndDispatch } from "../state";
+import * as Actions from "../state/actions";
+
+export async function loadToolboxCategoriesAsync() {
+    const { dispatch } = stateAndDispatch();
+
+    const regularCategories = await getToolboxCategories(false);
+    const advancedCategories = await getToolboxCategories(true);
+    const categories = (regularCategories ?? []).concat(advancedCategories ?? []);
+    if (categories.length === 0) {
+        return;
+    }
+
+    function shouldIncludeCategory(category: pxt.editor.ToolboxCategoryDefinition) {
+        return category && category.name && category.blocks?.length != 0;
+    }
+
+    // Create a map so categories can be looked up by their name.
+    const mappedCategories: pxt.Map<pxt.editor.ToolboxCategoryDefinition> = categories.reduce((map, category) => {
+        if (shouldIncludeCategory(category)) {
+            // Remove blocks with duplicate ids
+            const filteredBlocks = category.blocks?.filter((block, index, self) => {
+                return self.findIndex(b => b.blockId === block.blockId) === index;
+            });
+
+            map[category.name!] = { ...category, blocks: filteredBlocks };
```

Comment:
What if `category.name` is null? I know this is safe because I noticed `shouldIncludeCategory` checks it. Maybe put a comment to that effect here?